File: android/src/main/java/expo/modules/breathing/BreathingExerciseView.kt
package expo.modules.breathing

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.withFrameNanos
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.blur
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.scale
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Outline
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.LayoutDirection
import androidx.compose.ui.unit.dp
import expo.modules.orb.ParticlesView
import expo.modules.orb.RotationDirection
import expo.modules.orb.SimpleRotatingGlow
import expo.modules.orb.WavyBlobView
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.tan
import kotlin.random.Random

@Composable
fun BreathingExerciseView(
    config: BreathingConfiguration,
    modifier: Modifier = Modifier
) {
    // Frame-based animation state
    var frameTime by remember { mutableLongStateOf(System.nanoTime()) }
    var elapsedTime by remember { mutableStateOf(0.0) }

    // Continuous animation loop
    LaunchedEffect(Unit) {
        val startTime = System.nanoTime()
        while (true) {
            withFrameNanos { currentTime ->
                frameTime = currentTime
                elapsedTime = (currentTime - startTime) / 1_000_000_000.0
            }
        }
    }

    // Initialize wobble offsets if needed
    val state = BreathingSharedState
    if (state.wobbleOffsets.size != config.pointCount) {
        state.wobbleOffsets = MutableList(config.pointCount) { 0.0 }
        state.wobbleTargets = MutableList(config.pointCount) { 0.0 }
    }

    // Update animation state
    val animationState = updateAnimationState(frameTime, config)

    // Create blob shape for clipping
    val blobShape = remember(animationState.scale, animationState.wobbleOffsets, config.pointCount) {
        MorphingBlobShape(
            baseRadius = animationState.scale.toFloat() * 0.7f,
            pointCount = config.pointCount,
            offsets = animationState.wobbleOffsets
        )
    }

    Box(
        modifier = modifier.aspectRatio(1f),
        contentAlignment = Alignment.Center
    ) {
        // Progress ring (behind blob)
        if (config.showProgressRing && animationState.isActive) {
            ProgressRingView(
                progress = animationState.phaseProgress,
                color = config.progressRingColor,
                lineWidth = 4f,
                modifier = Modifier.fillMaxSize(0.95f)
            )
        }

        // Main blob with all effects
        Box(
            modifier = Modifier
                .fillMaxSize()
                .graphicsLayer { clip = false }
        ) {
            // Shadow
            if (config.showShadow) {
                BreathingShadow(
                    colors = config.blobColors,
                    scale = animationState.scale.toFloat(),
                    modifier = Modifier
                        .fillMaxSize()
                        .graphicsLayer { clip = false }
                )
            }

            // Main blob content clipped to morphing shape
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .clip(blobShape)
            ) {
                // Background gradient
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(
                            brush = Brush.verticalGradient(
                                colors = config.blobColors.reversed()
                            )
                        )
                )

                // Base depth glows
                BaseDepthGlows(
                    glowColor = config.glowColor,
                    elapsedTime = elapsedTime,
                    modifier = Modifier.fillMaxSize()
                )

                // Wavy blobs
                if (config.showWavyBlobs) {
                    WavyBlobLayer(
                        elapsedTime = elapsedTime,
                        modifier = Modifier.fillMaxSize()
                    )
                }

                // Core glow effects
                if (config.showGlowEffects) {
                    CoreGlowEffects(
                        glowColor = config.glowColor,
                        elapsedTime = elapsedTime,
                        modifier = Modifier
                            .fillMaxSize()
                            .padding(8.dp)
                    )
                }

                // Particles
                if (config.showParticles) {
                    ParticlesView(
                        color = config.particleColor,
                        particleCount = 15,
                        speedRange = 10f..20f,
                        sizeRange = 1f..3f,
                        opacityRange = 0.1f..0.4f,
                        elapsedTime = elapsedTime,
                        modifier = Modifier
                            .fillMaxSize()
                            .blur(1.dp)
                    )
                    ParticlesView(
                        color = config.particleColor,
                        particleCount = 10,
                        speedRange = 20f..35f,
                        sizeRange = 0.5f..2f,
                        opacityRange = 0.3f..0.7f,
                        elapsedTime = elapsedTime,
                        modifier = Modifier.fillMaxSize()
                    )
                }

                // Inner glow overlay
                InnerGlowOverlay(
                    modifier = Modifier.fillMaxSize()
                )
            }
        }

        // Text cue (on top)
        if (config.showTextCue && animationState.label.isNotEmpty()) {
            BreathingTextCue(
                text = animationState.label,
                color = config.textColor
            )
        }
    }
}

// Custom shape for clipping
private class MorphingBlobShape(
    private val baseRadius: Float,
    private val pointCount: Int,
    private val offsets: List<Double>
) : Shape {
    override fun createOutline(
        size: androidx.compose.ui.geometry.Size,
        layoutDirection: LayoutDirection,
        density: Density
    ): Outline {
        val path = Path()
        if (pointCount < 3) return Outline.Generic(path)

        val minDim = min(size.width, size.height)
        val center = Offset(size.width / 2f, size.height / 2f)
        val effectiveRadius = baseRadius * minDim / 2f

        // Generate points around the circle with offsets
        val points = mutableListOf<Offset>()
        for (i in 0 until pointCount) {
            val angle = (i.toDouble() / pointCount) * 2 * PI - PI / 2
            val offset = if (i < offsets.size) offsets[i] else 0.0
            val radius = effectiveRadius * (1.0 + offset).toFloat()

            points.add(
                Offset(
                    x = center.x + (cos(angle) * radius).toFloat(),
                    y = center.y + (sin(angle) * radius).toFloat()
                )
            )
        }

        // Tangent coefficient for smooth cubic bezier curves
        val tangentCoeff = ((4.0 / 3.0) * tan(PI / (2.0 * pointCount))).toFloat()

        // Start at first point
        path.moveTo(points[0].x, points[0].y)

        // Draw cubic bezier curves between each pair of points
        for (i in 0 until pointCount) {
            val current = points[i]
            val next = points[(i + 1) % pointCount]

            val currentAngle = (i.toDouble() / pointCount) * 2 * PI - PI / 2
            val nextAngle = (((i + 1) % pointCount).toDouble() / pointCount) * 2 * PI - PI / 2

            val currentOffset = if (i < offsets.size) offsets[i] else 0.0
            val currentRadius = effectiveRadius * (1.0 + currentOffset).toFloat()
            val currentTangentLength = currentRadius * tangentCoeff

            val nextOffset = if ((i + 1) % pointCount < offsets.size) offsets[(i + 1) % pointCount] else 0.0
            val nextRadius = effectiveRadius * (1.0 + nextOffset).toFloat()
            val nextTangentLength = nextRadius * tangentCoeff

            val control1 = Offset(
                x = current.x + (cos(currentAngle + PI / 2) * currentTangentLength).toFloat(),
                y = current.y + (sin(currentAngle + PI / 2) * currentTangentLength).toFloat()
            )

            val control2 = Offset(
                x = next.x + (cos(nextAngle - PI / 2) * nextTangentLength).toFloat(),
                y = next.y + (sin(nextAngle - PI / 2) * nextTangentLength).toFloat()
            )

            path.cubicTo(
                control1.x, control1.y,
                control2.x, control2.y,
                next.x, next.y
            )
        }

        path.close()
        return Outline.Generic(path)
    }
}

private data class AnimationState(
    val scale: Double,
    val wobbleOffsets: List<Double>,
    val phaseProgress: Double,
    val label: String,
    val isActive: Boolean
)

@Composable
private fun BaseDepthGlows(
    glowColor: Color,
    elapsedTime: Double,
    modifier: Modifier = Modifier
) {
    val speed = 18.0
    Box(modifier = modifier) {
        SimpleRotatingGlow(
            color = glowColor,
            rotationSpeed = speed * 0.75,
            direction = RotationDirection.COUNTER_CLOCKWISE,
            elapsedTime = elapsedTime,
            alpha = 0.5f,
            modifier = Modifier
                .fillMaxSize()
                .padding(4.dp)
                .blur(8.dp)
        )
        SimpleRotatingGlow(
            color = glowColor.copy(alpha = 0.5f),
            rotationSpeed = speed * 0.25,
            direction = RotationDirection.CLOCKWISE,
            elapsedTime = elapsedTime,
            alpha = 0.3f,
            modifier = Modifier
                .fillMaxSize()
                .padding(8.dp)
                .blur(4.dp)
        )
    }
}

@Composable
private fun CoreGlowEffects(
    glowColor: Color,
    elapsedTime: Double,
    modifier: Modifier = Modifier
) {
    val speed = 18.0
    val glowIntensity = 0.8
    Box(modifier = modifier) {
        SimpleRotatingGlow(
            color = glowColor,
            rotationSpeed = speed * 1.2,
            direction = RotationDirection.CLOCKWISE,
            elapsedTime = elapsedTime,
            alpha = (glowIntensity * 0.8).toFloat(),
            modifier = Modifier
                .fillMaxSize()
                .blur(12.dp)
        )
        SimpleRotatingGlow(
            color = glowColor,
            rotationSpeed = speed * 0.9,
            direction = RotationDirection.CLOCKWISE,
            elapsedTime = elapsedTime,
            alpha = (glowIntensity * 0.6).toFloat(),
            modifier = Modifier
                .fillMaxSize()
                .blur(8.dp)
        )
    }
}

@Composable
private fun WavyBlobLayer(
    elapsedTime: Double,
    modifier: Modifier = Modifier
) {
    val blobSpeed = 20.0
    Box(modifier = modifier) {
        WavyBlobView(
            color = Color.White.copy(alpha = 0.4f),
            loopDuration = 60.0 / blobSpeed * 1.75,
            elapsedTime = elapsedTime,
            modifier = Modifier
                .fillMaxSize()
                .scale(1.5f)
                .offset(y = 40.dp)
                .blur(2.dp)
        )
        WavyBlobView(
            color = Color.White.copy(alpha = 0.25f),
            loopDuration = 60.0 / blobSpeed * 2.25,
            elapsedTime = elapsedTime,
            modifier = Modifier
                .fillMaxSize()
                .scale(1.2f)
                .offset(y = (-40).dp)
                .graphicsLayer { rotationZ = 90f }
                .blur(2.dp)
        )
    }
}

@Composable
private fun InnerGlowOverlay(modifier: Modifier = Modifier) {
    Canvas(modifier = modifier) {
        val center = Offset(size.width / 2f, size.height / 2f)
        val radius = min(size.width, size.height) / 2f

        drawCircle(
            brush = Brush.verticalGradient(
                colors = listOf(Color.Transparent, Color.White.copy(alpha = 0.15f)),
            ),
            radius = radius,
            center = center
        )
    }
}

@Composable
private fun BreathingShadow(
    colors: List<Color>,
    scale: Float,
    modifier: Modifier = Modifier
) {
    Box(modifier = modifier) {
        // Soft outer shadow
        Canvas(
            modifier = Modifier
                .fillMaxSize()
                .offset(y = 8.dp)
                .blur(24.dp)
                .graphicsLayer { alpha = 0.3f }
        ) {
            drawCircle(
                brush = Brush.verticalGradient(colors.reversed()),
                radius = size.minDimension / 2f * scale * 0.7f
            )
        }
        // Closer shadow
        Canvas(
            modifier = Modifier
                .fillMaxSize()
                .offset(y = 4.dp)
                .blur(12.dp)
                .graphicsLayer { alpha = 0.5f }
        ) {
            drawCircle(
                brush = Brush.verticalGradient(colors.reversed()),
                radius = size.minDimension / 2f * scale * 0.7f
            )
        }
    }
}

private fun easeOutCubic(t: Double): Double {
    val t1 = t - 1
    return t1 * t1 * t1 + 1
}

private fun updateAnimationState(
    frameTime: Long,
    config: BreathingConfiguration
): AnimationState {
    val state = BreathingSharedState

    // Always update wobble animation
    updateWobble(frameTime, config)

    return when (state.state) {
        BreathingExerciseState.STOPPED, BreathingExerciseState.COMPLETE -> {
            AnimationState(
                scale = 1.0,
                wobbleOffsets = state.wobbleOffsets.toList(),
                phaseProgress = 0.0,
                label = "",
                isActive = false
            )
        }

        BreathingExerciseState.PAUSED -> {
            AnimationState(
                scale = state.currentScale,
                wobbleOffsets = state.wobbleOffsets.toList(),
                phaseProgress = state.phaseProgress,
                label = state.currentLabel,
                isActive = true
            )
        }

        BreathingExerciseState.RUNNING -> {
            // Update phase timing
            updatePhaseState(frameTime)

            // Calculate scale and progress based on phase
            val currentPhaseConfig = if (state.phases.isEmpty()) null else state.phases[state.currentPhaseIndex]
            val scale: Double
            val ringProgress: Double

            if (currentPhaseConfig != null) {
                when (currentPhaseConfig.phase) {
                    BreathPhase.INHALE -> {
                        val easedProgress = easeOutCubic(state.phaseProgress)
                        scale = state.startScale + (state.targetScale - state.startScale) * easedProgress
                        ringProgress = easedProgress
                    }
                    BreathPhase.EXHALE -> {
                        val easedProgress = easeOutCubic(state.phaseProgress)
                        scale = state.startScale + (state.targetScale - state.startScale) * easedProgress
                        ringProgress = 1.0 - easedProgress
                    }
                    BreathPhase.HOLD_IN -> {
                        scale = state.targetScale
                        ringProgress = 1.0
                    }
                    BreathPhase.HOLD_OUT -> {
                        scale = state.targetScale
                        ringProgress = 0.0
                    }
                    BreathPhase.IDLE -> {
                        scale = state.targetScale
                        ringProgress = 0.0
                    }
                }
            } else {
                scale = 1.0
                ringProgress = 0.0
            }

            state.currentScale = scale

            AnimationState(
                scale = scale,
                wobbleOffsets = state.wobbleOffsets.toList(),
                phaseProgress = ringProgress,
                label = state.currentLabel,
                isActive = true
            )
        }
    }
}

private fun updatePhaseState(frameTime: Long) {
    val state = BreathingSharedState
    if (state.phases.isEmpty()) return

    val currentPhaseConfig = state.phases[state.currentPhaseIndex]
    val elapsed = (frameTime - state.phaseStartTime) / 1_000_000_000.0
    val duration = currentPhaseConfig.duration

    // Calculate progress through current phase
    state.phaseProgress = min(1.0, elapsed / duration)

    // Check if phase is complete
    if (elapsed >= duration) {
        // Move to next phase
        state.currentPhaseIndex = (state.currentPhaseIndex + 1) % state.phases.size

        // Check if we completed a cycle
        if (state.currentPhaseIndex == 0) {
            state.currentCycle += 1

            // Check if exercise is complete
            val totalCycles = state.totalCycles
            if (totalCycles != null && state.currentCycle >= totalCycles) {
                state.state = BreathingExerciseState.COMPLETE
                state.totalDuration = (frameTime - state.exerciseStartTime) / 1_000_000_000.0
                state.onExerciseComplete?.invoke(state.currentCycle, state.totalDuration)
                return
            }
        }

        // Start new phase
        state.phaseStartTime = frameTime
        val newPhaseConfig = state.phases[state.currentPhaseIndex]
        state.currentPhase = newPhaseConfig.phase
        state.currentLabel = newPhaseConfig.label
        state.startScale = state.currentScale
        state.targetScale = newPhaseConfig.targetScale
        state.phaseProgress = 0.0

        // Update wobble intensity based on phase
        state.wobbleIntensity = when (newPhaseConfig.phase) {
            BreathPhase.INHALE, BreathPhase.EXHALE -> 1.0
            BreathPhase.HOLD_IN, BreathPhase.HOLD_OUT -> 0.3
            BreathPhase.IDLE -> 0.5
        }

        // Fire phase change callback
        state.onPhaseChange?.invoke(
            newPhaseConfig.phase,
            newPhaseConfig.label,
            state.currentPhaseIndex,
            state.currentCycle
        )
    }
}

private fun updateWobble(frameTime: Long, config: BreathingConfiguration) {
    val state = BreathingSharedState

    // Ensure arrays are sized correctly
    if (state.wobbleOffsets.size != config.pointCount) {
        state.wobbleOffsets = MutableList(config.pointCount) { 0.0 }
        state.wobbleTargets = MutableList(config.pointCount) { 0.0 }
    }

    // Check if we need new random targets (slower, more organic)
    val timeSinceLastTarget = (frameTime - state.lastWobbleTargetUpdate) / 1_000_000_000.0
    if (timeSinceLastTarget > 1.8 || state.wobbleTargets.all { it == 0.0 }) {
        state.lastWobbleTargetUpdate = frameTime
        val intensity = state.wobbleIntensity * config.wobbleIntensity
        val maxOffset = 0.18 * intensity

        for (i in 0 until config.pointCount) {
            val phase = i.toDouble() / config.pointCount * 2.0 * PI
            val wave1 = sin(phase * 2 + Random.nextDouble()) * 0.6
            val wave2 = cos(phase * 3 + Random.nextDouble()) * 0.4
            val baseOffset = Random.nextDouble(-maxOffset, maxOffset)
            state.wobbleTargets[i] = baseOffset * (1.0 + wave1 + wave2) * 0.5
        }
    }

    // Smoothly interpolate offsets toward targets
    val dt = (frameTime - state.lastWobbleUpdate) / 1_000_000_000.0
    state.lastWobbleUpdate = frameTime
    val interpolationSpeed = 1.8
    val factor = min(1.0, dt * interpolationSpeed)

    for (i in 0 until min(state.wobbleOffsets.size, state.wobbleTargets.size)) {
        state.wobbleOffsets[i] = state.wobbleOffsets[i] + (state.wobbleTargets[i] - state.wobbleOffsets[i]) * factor
    }
}

File: ios/Breathing/BreathingExerciseView.swift
import SwiftUI

public struct BreathingExerciseView: View {
    private let config: BreathingConfiguration

    public init(configuration: BreathingConfiguration = BreathingConfiguration()) {
        self.config = configuration
    }

    public var body: some View {
        TimelineView(.animation) { timeline in
            GeometryReader { geometry in
                let size = min(geometry.size.width, geometry.size.height)
                let elapsedTime = timeline.date.timeIntervalSinceReferenceDate

                // Update all animation state
                let animState = updateAllAnimations(at: timeline.date, pointCount: config.pointCount)

                ZStack {
                    // Base gradient background layer
                    LinearGradient(
                        colors: config.blobColors,
                        startPoint: .bottom,
                        endPoint: .top
                    )

                    // Base depth glows - creates depth with rotating glow effects
                    baseDepthGlows(size: size, elapsedTime: elapsedTime)

                    // Wavy blobs - adds organic movement with flowing blob shapes
                    if config.showWavyBlobs {
                        wavyBlob(size: size, elapsedTime: elapsedTime)
                        wavyBlobTwo(size: size, elapsedTime: elapsedTime)
                    }

                    // Core glow effects - adds bright, energetic core glow animations
                    if config.showGlowEffects {
                        coreGlowEffects(size: size, elapsedTime: elapsedTime)
                    }

                    // Particles - overlays floating particle effects for additional dynamism
                    if config.showParticles {
                        particleView
                            .frame(maxWidth: size, maxHeight: size)
                    }

                    // Inner animated blob
                    if config.showInnerBlob {
                        InnerWavyBlob(
                            center: CGPoint(x: size / 2, y: size / 2),
                            baseRadius: animState.scale * 0.7 * size / 2 * 0.55,
                            color: config.innerBlobColor,
                            elapsedTime: elapsedTime
                        )
                        .blur(radius: 10)
                        .blendMode(.plusLighter)
                    }
                }
                // Inner glows overlay for depth
                .overlay {
                    realisticInnerGlows
                }
                // Mask with morphing blob shape
                .mask {
                    MorphingBlobShape(
                        center: CGPoint(x: size / 2, y: size / 2),
                        baseRadius: animState.scale * 0.7 * size / 2,
                        pointCount: config.pointCount,
                        offsets: animState.wobbleOffsets
                    )
                }
                .frame(width: size, height: size)
                .background {
                    // Progress ring (behind blob, outside mask)
                    if config.showProgressRing && animState.isActive {
                        ProgressRingView(
                            progress: animState.phaseProgress,
                            color: config.progressRingColor,
                            lineWidth: 4,
                            size: size * 0.95
                        )
                    }
                }
                .overlay {
                    // Text cue
                    if config.showTextCue && !animState.label.isEmpty {
                        BreathingTextCue(
                            text: animState.label,
                            color: config.textColor
                        )
                    }
                }
                .modifier(
                    BreathingShadowModifier(
                        colors: config.showShadow ? config.blobColors : [.clear],
                        radius: size * 0.06,
                        scale: animState.scale
                    )
                )
            }
            .aspectRatio(1, contentMode: .fit)
        }
    }

    // MARK: - Visual Effects (from OrbView)

    private var particleView: some View {
        ZStack {
            BreathingParticlesView(
                color: config.particleColor,
                particleCount: 15,
                speedRange: 10...20,
                sizeRange: 1...3,
                opacityRange: 0.1...0.4
            )
            .blur(radius: 1)

            BreathingParticlesView(
                color: config.particleColor,
                particleCount: 10,
                speedRange: 20...35,
                sizeRange: 0.5...2,
                opacityRange: 0.3...0.7
            )
        }
        .blendMode(.plusLighter)
    }

    private func wavyBlob(size: CGFloat, elapsedTime: Double) -> some View {
        let blobSpeed: Double = 20

        return RotatingGlowView(
            color: .white.opacity(0.75),
            rotationSpeed: blobSpeed * 1.5,
            direction: .clockwise,
            elapsedTime: elapsedTime
        )
        .mask {
            WavyBlobView(color: .white, loopDuration: 60 / blobSpeed * 1.75)
                .frame(maxWidth: size * 1.875)
                .offset(x: 0, y: size * 0.31)
        }
        .blur(radius: 1)
        .blendMode(.plusLighter)
    }

    private func wavyBlobTwo(size: CGFloat, elapsedTime: Double) -> some View {
        let blobSpeed: Double = 20

        return RotatingGlowView(
            color: .white,
            rotationSpeed: blobSpeed * 0.75,
            direction: .counterClockwise,
            elapsedTime: elapsedTime
        )
        .mask {
            WavyBlobView(color: .white, loopDuration: 60 / blobSpeed * 2.25)
                .frame(maxWidth: size * 1.25)
                .rotationEffect(.degrees(90))
                .offset(x: 0, y: size * -0.31)
        }
        .opacity(0.5)
        .blur(radius: 1)
        .blendMode(.plusLighter)
    }

    private func coreGlowEffects(size: CGFloat, elapsedTime: Double) -> some View {
        let speed: Double = 18.0
        let glowIntensity: Double = 0.8

        return ZStack {
            RotatingGlowView(
                color: config.glowColor,
                rotationSpeed: speed * 1.2,
                direction: .clockwise,
                elapsedTime: elapsedTime
            )
            .blur(radius: size * 0.08)
            .opacity(glowIntensity)

            RotatingGlowView(
                color: config.glowColor,
                rotationSpeed: speed * 0.9,
                direction: .clockwise,
                elapsedTime: elapsedTime
            )
            .blur(radius: size * 0.06)
            .opacity(glowIntensity)
            .blendMode(.plusLighter)
        }
        .padding(size * 0.08)
    }

    private func baseDepthGlows(size: CGFloat, elapsedTime: Double) -> some View {
        let speed: Double = 18.0

        return ZStack {
            // Outer glow
            RotatingGlowView(
                color: config.glowColor,
                rotationSpeed: speed * 0.75,
                direction: .counterClockwise,
                elapsedTime: elapsedTime
            )
            .padding(size * 0.03)
            .blur(radius: size * 0.06)
            .rotationEffect(.degrees(180))
            .blendMode(.destinationOver)

            // Outer ring
            RotatingGlowView(
                color: config.glowColor.opacity(0.5),
                rotationSpeed: speed * 0.25,
                direction: .clockwise,
                elapsedTime: elapsedTime
            )
            .frame(maxWidth: size * 0.94)
            .rotationEffect(.degrees(180))
            .padding(8)
            .blur(radius: size * 0.032)
        }
    }

    private var orbOutlineColor: LinearGradient {
        LinearGradient(
            colors: [.white, .clear],
            startPoint: .bottom,
            endPoint: .top
        )
    }

    private var realisticInnerGlows: some View {
        ZStack {
            // Outer stroke with heavy blur
            Circle()
                .stroke(orbOutlineColor, lineWidth: 8)
                .blur(radius: 32)
                .blendMode(.plusLighter)

            // Inner stroke with light blur
            Circle()
                .stroke(orbOutlineColor, lineWidth: 4)
                .blur(radius: 12)
                .blendMode(.plusLighter)

            Circle()
                .stroke(orbOutlineColor, lineWidth: 1)
                .blur(radius: 4)
                .blendMode(.plusLighter)
        }
        .padding(1)
    }

    private struct AnimState {
        let scale: Double
        let wobbleOffsets: [Double]
        let phaseProgress: Double
        let label: String
        let isActive: Bool
    }

    private func updateAllAnimations(at date: Date, pointCount: Int) -> AnimState {
        let state = BreathingSharedState.shared

        // Always update wobble animation (organic movement even when idle)
        updateWobble(at: date, pointCount: pointCount)

        // Handle breathing state
        switch state.state {
        case .stopped, .complete:
            return AnimState(
                scale: 1.0,
                wobbleOffsets: state.wobbleOffsets,
                phaseProgress: 0,
                label: "",
                isActive: false
            )

        case .paused:
            return AnimState(
                scale: state.currentScale,
                wobbleOffsets: state.wobbleOffsets,
                phaseProgress: state.phaseProgress,
                label: state.currentLabel,
                isActive: true
            )

        case .running:
            // Update phase timing
            updatePhaseState(at: date)

            // Calculate scale and progress based on phase
            let currentPhaseConfig = state.phases.isEmpty ? nil : state.phases[state.currentPhaseIndex]
            let scale: Double
            let ringProgress: Double

            if let phase = currentPhaseConfig?.phase {
                switch phase {
                case .inhale:
                    // Ease-out cubic for smooth deceleration toward target
                    let easedProgress = easeOutCubic(state.phaseProgress)
                    scale = state.startScale + (state.targetScale - state.startScale) * easedProgress
                    ringProgress = easedProgress  // 0 -> 1
                case .exhale:
                    let easedProgress = easeOutCubic(state.phaseProgress)
                    scale = state.startScale + (state.targetScale - state.startScale) * easedProgress
                    ringProgress = 1.0 - easedProgress  // 1 -> 0
                case .holdIn:
                    // Hold at expanded scale, ring stays full
                    scale = state.targetScale
                    ringProgress = 1.0
                case .holdOut:
                    // Hold at contracted scale, ring stays empty
                    scale = state.targetScale
                    ringProgress = 0.0
                case .idle:
                    scale = state.targetScale
                    ringProgress = 0.0
                }
            } else {
                scale = 1.0
                ringProgress = 0.0
            }

            state.currentScale = scale

            return AnimState(
                scale: scale,
                wobbleOffsets: state.wobbleOffsets,
                phaseProgress: ringProgress,
                label: state.currentLabel,
                isActive: true
            )
        }
    }

    private func updatePhaseState(at date: Date) {
        let state = BreathingSharedState.shared
        guard !state.phases.isEmpty else { return }

        let currentPhaseConfig = state.phases[state.currentPhaseIndex]
        let elapsed = date.timeIntervalSince(state.phaseStartTime)
        let duration = currentPhaseConfig.duration

        // Calculate progress through current phase
        state.phaseProgress = min(1.0, elapsed / duration)

        // Check if phase is complete
        if elapsed >= duration {
            // Move to next phase
            state.currentPhaseIndex = (state.currentPhaseIndex + 1) % state.phases.count

            // Check if we completed a cycle
            if state.currentPhaseIndex == 0 {
                state.currentCycle += 1

                // Check if exercise is complete
                if let totalCycles = state.totalCycles, state.currentCycle >= totalCycles {
                    state.state = .complete
                    state.totalDuration = date.timeIntervalSince(state.exerciseStartTime)
                    state.onExerciseComplete?(state.currentCycle, state.totalDuration)
                    return
                }
            }

            // Start new phase
            state.phaseStartTime = date
            let newPhaseConfig = state.phases[state.currentPhaseIndex]
            state.currentPhase = newPhaseConfig.phase
            state.currentLabel = newPhaseConfig.label
            state.startScale = state.currentScale  // Remember where we're starting from
            state.targetScale = newPhaseConfig.targetScale
            state.phaseProgress = 0

            // Update wobble intensity based on phase
            switch newPhaseConfig.phase {
            case .inhale, .exhale:
                state.wobbleIntensity = 1.0
            case .holdIn, .holdOut:
                state.wobbleIntensity = 0.3
            case .idle:
                state.wobbleIntensity = 0.5
            }

            // Fire phase change callback
            state.onPhaseChange?(
                newPhaseConfig.phase,
                newPhaseConfig.label,
                state.currentPhaseIndex,
                state.currentCycle
            )
        }
    }

    private func updateWobble(at date: Date, pointCount: Int) {
        let state = BreathingSharedState.shared

        // Initialize arrays if needed
        if state.wobbleOffsets.count != pointCount {
            state.wobbleOffsets = (0..<pointCount).map { _ in 0.0 }
            state.wobbleTargets = (0..<pointCount).map { _ in 0.0 }
            state.lastWobbleUpdate = date
            state.lastWobbleTargetUpdate = date
        }

        let now = date
        let dt = now.timeIntervalSince(state.lastWobbleUpdate)
        state.lastWobbleUpdate = now

        // Generate new random targets periodically (slower, more organic)
        let timeSinceTargetUpdate = now.timeIntervalSince(state.lastWobbleTargetUpdate)
        if timeSinceTargetUpdate > 1.8 {
            state.lastWobbleTargetUpdate = now

            let intensity = state.wobbleIntensity * config.wobbleIntensity
            let maxOffset = 0.18 * intensity

            state.wobbleTargets = (0..<pointCount).map { i in
                let phase = Double(i) / Double(pointCount) * 2.0 * .pi
                let wave1 = sin(phase * 2 + Double.random(in: 0...1)) * 0.6
                let wave2 = cos(phase * 3 + Double.random(in: 0...1)) * 0.4
                let baseOffset = Double.random(in: -maxOffset...maxOffset)
                return baseOffset * (1.0 + wave1 + wave2) * 0.5
            }
        }

        // Smoothly interpolate current offsets toward targets
        let interpolationSpeed = 1.8
        let factor = min(1.0, dt * interpolationSpeed)

        for i in 0..<pointCount {
            let current = state.wobbleOffsets[i]
            let target = state.wobbleTargets[i]
            state.wobbleOffsets[i] = current + (target - current) * factor
        }
    }
}

struct BreathingShadowModifier: ViewModifier {
    let colors: [Color]
    let radius: CGFloat
    let scale: Double

    func body(content: Content) -> some View {
        content
            .shadow(color: colors.first?.opacity(0.4) ?? .clear, radius: radius * scale, x: 0, y: radius * 0.3 * scale)
            .shadow(color: colors.first?.opacity(0.2) ?? .clear, radius: radius * 2 * scale, x: 0, y: radius * 0.5 * scale)
    }
}

// Easing functions
private func easeOutCubic(_ t: Double) -> Double {
    let t1 = t - 1
    return t1 * t1 * t1 + 1
}

private func easeInOutCubic(_ t: Double) -> Double {
    if t < 0.5 {
        return 4 * t * t * t
    } else {
        let t1 = -2 * t + 2
        return 1 - (t1 * t1 * t1) / 2
    }
}
